aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/models/AWSPinpointUserProfile.java

```diff
@@ -0,0 +1,189 @@
+ * relevant to Pinpoint only.
+ */
+public final class AWSPinpointUserProfile extends UserProfile {
+    private final AnalyticsProperties userAttributes;
+
+    /**
+     * Constructor that mirrors the parent class.
+     * @param builder An instance of the builder with the desired properties set.
+     */
+    protected AWSPinpointUserProfile(@NonNull Builder builder) {
+        super(UserProfile.builder()
+                .name(builder.name)
+                .email(builder.email)
+                .plan(builder.plan)
+                .location(builder.location)
+                .customProperties(builder.customProperties));
+        this.userAttributes = builder.userAttributes;
+    }
+
+    /**
+     * Gets all the available user attributes.
+     * @return The user's attributes.
+     */
+    @Nullable
+    public AnalyticsProperties getUserAttributes() {
+        return userAttributes;
+    }
+
+    /**
+     * Begins construction of an {@link AWSPinpointUserProfile} using a builder pattern.
+     * @return An {@link AWSPinpointUserProfile.Builder} instance
+     */
+    @NonNull
+    public static Builder builder() {
+        return new AWSPinpointUserProfile.Builder();
+    }
+
+    @Override
+    public boolean equals(@Nullable Object object) {
+        boolean isEquals = super.equals(object);
+//        if (this == object) {
+//            return true;
+//        }
+//        if (object == null || !(object instanceof AWSPinpointUserProfile)) {
+//            return false;
+//        }
+        AWSPinpointUserProfile that = (AWSPinpointUserProfile) object;
+        return isEquals && ObjectsCompat.equals(userAttributes, that.userAttributes);
+    }
+
+    @Override
+    public int hashCode() {
+        return 31 * super.hashCode() + (userAttributes != null ? userAttributes.hashCode() : 0);
+    }
+
+    @NonNull
+    @Override
+    public String toString() {
+        return "UserProfile{" +
+                "name='" + getName() + '\'' +
+                ", email='" + getEmail() + '\'' +
+                ", plan='" + getPlan() + '\'' +
+                ", location=" + getLocation() +
+                ", customProperties=" + getCustomProperties() +
+                ", userProperties=" + userAttributes +
+                '}';
+    }
+
+    /**
+     * Builder for the {@link AWSPinpointUserProfile} class.
+     */
+    public static final class Builder extends UserProfile.Builder {
+        private String name;
+        private String email;
+        private String plan;
+        private Location location;
+        private AnalyticsProperties customProperties;
+        private AnalyticsProperties userAttributes;
+
+        /**
+         * Configures the name to be used in the next-built UserProfile.
+         * @param name User's name
+         * @return Current builder instance, for method chaining
+         */
+        @NonNull
+        public Builder name(@NonNull final String name) {
```

Comment:
We should be able to avoid having to reimplement all of these methods from `UserProfile.Builder` by making `UserProfile.Builder` generic.  That is, each builder method should return an object of type `<B>`, instead of a concrete `UserProfile.Builder`.  See `StorageOptions.Builder` / `StorageUploadOptions.Builder` for an example of this.

Reply:
Just updated this. Not sure why, but every time I look at this Builder pattern, it takes me a half an hour to wrap my head around it 🤣